File: plugins/collector/event/apps/bigdata/hbase.go
package bigdata

import (
	"collector/cache/process"
	"collector/event/apps"
	"path/filepath"
	"regexp"
	"strings"
)

// Hadoop database, put this in bigdata temporary
type Hbase struct {
	version    string
	reg        *regexp.Regexp
	regDir     *regexp.Regexp
	regVersion *regexp.Regexp
}

func (Hbase) Name() string { return "hbase" }

func (Hbase) Type() string { return "bigdata" }

func (h Hbase) Version() string { return h.version }

func (h *Hbase) Match(p *process.Process) bool {
	if p.Name != "java" {
		return false
	}
	if h.reg == nil {
		h.reg = regexp.MustCompile(` org\.apache\.hadoop\.hbase\.master`)
	}
	if s := h.reg.FindString(p.Argv); s == "" {
		return false
	}
	return true
}

func (h *Hbase) Run(p *process.Process) (m map[string]string, err error) {
	m = make(map[string]string)
	if h.regDir == nil {
		h.regDir = regexp.MustCompile(`hbase.home.dir=[a-zA-Z0-9\/-_]+`)
	}
	if h.regVersion == nil {
		h.regVersion = regexp.MustCompile(`HBase (\d+\.)+\d`)
	}
	var rootPath string
	if s := h.regDir.FindString(p.Argv); s != "" {
		rootPath = strings.TrimPrefix(s, "hbase.home.dir=")
	} else {
		return
	}
	if result, err := apps.ExecuteWithName(p, filepath.Join(rootPath, "hbase"), "version"); err == nil {
		if s := h.regVersion.FindString(result); s != "" {
			h.version = strings.TrimPrefix(s, "HBase ")
		}
	}
	return
}

func init() { apps.Regist(&Hbase{}) }

File: plugins/collector/src/event/apps/bigdata/hbase.rs
use crate::event::apps::{find_version, jar_names_for_pid, AppProc, IApp};
use std::collections::HashMap;

pub struct Hbase {
    version: String,
}

impl Hbase {
    pub fn new() -> Self { Hbase { version: String::new() } }
}

impl IApp for Hbase {
    fn name(&self)     -> &'static str { "hbase" }
    fn app_type(&self) -> &'static str { "bigdata" }
    fn version(&self)  -> &str         { &self.version }

    fn matches(&mut self, p: &AppProc) -> bool {
        p.name == "java" && p.argv.contains("org.apache.hadoop.hbase.master")
    }

    fn run(&mut self, p: &AppProc) -> anyhow::Result<HashMap<String, String>> {
        self.version.clear();
        // "hbase-server-2.4.13.jar"
        for jar in jar_names_for_pid(p.pid) {
            if jar.starts_with("hbase-server-") && jar.ends_with(".jar") {
                let mid = jar.trim_start_matches("hbase-server-").trim_end_matches(".jar");
                let ver = find_version(mid);
                if !ver.is_empty() { self.version = ver; break; }
            }
        }
        Ok(HashMap::new())
    }
}
